File: receiver-windows/src/OpenAudioLink/Protocol/PacketReader.cs
using System;
using System.IO;

namespace OpenAudioLink.Protocol
{
    public static class PacketReader
    {
        public static byte[] ReadPacket(Stream stream)
        {
            byte[] header = ReadExact(stream, ProtocolConstants.HeaderSize);
            uint payloadLength = PacketParser.ReadUInt32(header, 20);
            if (payloadLength > ProtocolConstants.MaxPacketSize)
            {
                throw new PacketParseException("Payload is too large.");
            }

            byte[] packet = new byte[ProtocolConstants.HeaderSize + (int)payloadLength];
            Buffer.BlockCopy(header, 0, packet, 0, header.Length);
            byte[] payload = ReadExact(stream, (int)payloadLength);
            Buffer.BlockCopy(payload, 0, packet, ProtocolConstants.HeaderSize, payload.Length);
            PacketParser.ParseHeader(packet);
            return packet;
        }

        private static byte[] ReadExact(Stream stream, int length)
        {
            byte[] buffer = new byte[length];
            int offset = 0;
            while (offset < length)
            {
                int read = stream.Read(buffer, offset, length - offset);
                if (read == 0)
                {
                    throw new EndOfStreamException("Connection closed while reading packet.");
                }

                offset += read;
            }

            return buffer;
        }
    }
}

File: sender-android/app/src/main/java/com/openaudiolink/protocol/PacketReader.kt
package com.openaudiolink.protocol

import java.io.EOFException
import java.io.InputStream

data class Packet(val header: PacketHeader, val payload: ByteArray)

object PacketReader {
    fun read(input: InputStream): Packet {
        val headerBytes = input.readFully(ProtocolConstants.HeaderSize)
        val payloadLength = PacketParser.readUInt32(headerBytes, 20)
        if (payloadLength > ProtocolConstants.MaxPacketSize) PacketParser.parseHeader(headerBytes)
        val payload = input.readFully(payloadLength.toInt())
        val packet = headerBytes + payload
        return Packet(PacketParser.parseHeader(packet), payload)
    }

    private fun InputStream.readFully(size: Int): ByteArray {
        val buffer = ByteArray(size)
        var offset = 0
        while (offset < size) {
            val read = read(buffer, offset, size - offset)
            if (read < 0) throw EOFException("Premature end of stream.")
            offset += read
        }
        return buffer
    }
}
